### PR TITLE
feat(encoding): write frame content size in encoder output

### LINKED ISSUE #16: feat: frame content size in encoder output
## Summary

The encoder currently sets `frame_content_size: None` in the frame header (`frame_compressor.rs:145`). The C reference defaults to writing content size when known — this enables decoders to pre-allocate output buffers, significantly improving decompression speed.

## C reference behavior

- `ZSTD_c_contentSizeFlag` defaults to **1** (enabled)
- When pledged source size is known, it's written in 1/2/4/8 bytes depending on value
- Single-segment frames (content fits in window) use content size instead of window descriptor — saves 1 byte per frame

## Current Rust state

- `frame_header.rs:74` — FCS field always `None`
- Header always writes window descriptor even when content size would suffice
- `frame_compressor.rs:144-148` — `single_segment: false` always

## What needs to be implemented

1. **Content size tracking** — count total bytes read during compression
2. **Two-pass or buffered approach** — either buffer all input to know size upfront, or write placeholder and seek back
3. **Single-segment optimization** — when content ≤ window, set `single_segment: true` and omit window descriptor
4. **FCS encoding** — 1/2/4/8 byte encoding based on value

## Acceptance criteria
- [ ] Frame content size written when source size is known
- [ ] Single-segment flag set when applicable
- [ ] C zstd decompresses correctly (validates FCS)
- [ ] Decompressor can pre-allocate buffer from FCS

## Time estimate
1d

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- `FrameCompressor` now automatically writes Frame Content Size (FCS) in the frame header by buffering compressed blocks and deferring the header until total input size is known
- `StreamingEncoder` gains `set_pledged_content_size()` API for callers who know input size upfront; `finish()` verifies the pledge matches bytes written
- Fix FCS descriptor flag mapping bug (`8 => 3`, was `3 => 8` which would panic on 8-byte FCS values)
- Add `find_fcs_field_size()` with correct +256 offset range for 2-byte FCS fields (256–65791)
- Always include window descriptor alongside FCS (no `single_segment`) to avoid C zstd incompatibility with compressed blocks whose encoder window exceeds the declared content size

## Technical Details

The C reference defaults `ZSTD_c_contentSizeFlag` to 1 (enabled) and writes FCS whenever the pledged source size is known. This enables decoders to pre-allocate output buffers, significantly improving decompression speed.

**FrameCompressor**: blocks are accumulated in an intermediate buffer instead of being flushed to the drain per-block. After all input is consumed, the correct header (with FCS) is written first, then the buffered blocks. This trades O(compressed_size) memory for automatic FCS detection without requiring callers to pledge.

**StreamingEncoder**: since the header must be written before data flows, FCS requires an explicit pledge via `set_pledged_content_size()`. Without a pledge, the header omits FCS (matching previous behavior).

**No `single_segment`**: the C encoder sets `single_segment` when `windowSize >= pledgedSrcSize`, which makes the decoder use FCS as window size. Our compressed blocks are encoded against the matcher's actual window (128KB+), so setting `single_segment` with a small FCS would give the decoder a window too small for the block format. We always include the window descriptor instead (costs 1 byte per frame).

## Test Plan

- [x] 185 lib tests pass (including 1000-iteration roundtrip suites)
- [x] 15 cross-validation tests (Rust encoder ↔ C decoder and vice versa)
- [x] New `fcs_header_written_and_c_zstd_compatible` — all 5 levels × 4 input sizes verified against C zstd
- [x] New `pledged_content_size_written_in_header` — FCS roundtrip via StreamingEncoder
- [x] New `pledged_content_size_mismatch_returns_error` — pledge verification
- [x] New `pledged_content_size_after_write_returns_error` — API misuse guard
- [x] New `pledged_content_size_c_zstd_compatible` — streaming with pledge decoded by C zstd
- [x] New `no_pledged_size_omits_fcs_from_header` — backward compatibility
- [x] New `find_fcs_field_size` unit tests for both single-segment and non-single-segment paths
- [x] Clippy clean, doc-tests pass

Closes #16

## Summary by CodeRabbit

* **New Features**
  * Non-streaming compressor now emits Frame Content Size (FCS) automatically; streaming encoder adds an API to pledge an upfront frame size and enforces it.

* **Bug Fixes**
  * Corrected FCS field sizing/serialization for external decoder compatibility.
  * Improved empty-input handling and stricter pledge/size validation during streaming writes and finish.

* **Documentation**
  * README updated to mark FCS implemented and note streamer pledge requirement.

* **Tests**
  * Added tests for pledged-size behavior, round-trip decoding, compatibility, and edge cases.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ Complete RFC 8878 implementation. Performance: ~1.4-3.5x slower than C zstd depe
 - [x] Better (roughly level 7)
 - [x] Best (roughly level 11)
 - [x] Checksums
+- [x] Frame Content Size (enables decoder pre-allocation)
```
[the remark this answers is not in the document]
Comment (polaz) [reply README.md]: @coderabbitai Fixed — clarified both FCS code paths in the README line.

File: zstd/src/encoding/streaming_encoder.rs [hunk 2]
```diff
@@ -68,11 +70,31 @@ impl<W: Write, M: Matcher> StreamingEncoder<W, M> {
...
+    /// calling it after the frame header has already been emitted returns an
+    /// error.
+    pub fn set_pledged_content_size(&mut self, size: u64) -> Result<(), Error> {
+        if self.frame_started {
+            return Err(invalid_input_error(
+                "pledged content size must be set before the first write",
+            ));
+        }
+        self.pledged_content_size = Some(size);
+        Ok(())
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/streaming_encoder.rs]: @coderabbitai Fixed — added ensure_open() call at the start of set_pledged_content_size().
